cleverhans/picklable_model.py

```diff
@@ -496,5 +509,48 @@ def set_input_shape(self, shape):
     def get_params(self):
         return []
 
-    def fprop(self, x):
+    def fprop(self, x, **kwargs):
         return tf.map_fn(lambda ex: tf.image.per_image_standardization(ex), x)
+
+
+class Dropout(Layer):
+    """Dropout layer.
+
+    By default, is a no-op. Activate it during training using the kwargs
+    of MLP.fprop.
+    """
+
+    def __init__(self, include_prob=0.5, **kwargs):
+        super(Dropout, self).__init__(**kwargs)
+        self.include_prob = include_prob
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_params(self):
+        return []
+
+    def fprop(self, x, dropout=False, dropout_dict=None, **kwargs):
+        """
+        Forward propagation as either no-op or dropping random units.
+        :param x: The input to the layer
+        :param dropout: bool specifying whether to drop units
+        :param dropout_dict: dict
+            This dictionary is usually not needed.
+            In rare cases, generally for research purposes, this dictionary
+            makes it possible to run forward propagation with a different
+            dropout include probability.
+            This dictionary should be passed as a named argument to the MLP
+            class, which will then pass it to *all* layers' fprop methods.
+            Other layers will just receive this as an ignored kwargs entry.
+            Each dropout layer looks up its own name in this dictionary
+            to read out its include probability.
+        """
+        include_prob = self.include_prob
+        if dropout_dict is not None:
+            assert dropout
```

Comment:
Just making sure I understand the intended use here. Does that mean that all layers must have `dropout` set to `True` even if their `include_prob` is 0 when `dropout_dict` is not `None`?  Does that also mean that an include probability must be specified for all dropout layers in the `dropout_dict`?

Reply:
The default use case is that you never specify `include_prob` anywhere. During evaluation, you don't do anything special regarding dropout, and nothing gets dropped. During training, you pass "dropout=True" to make units get randomly dropped. If you've done nothing else, include_prob defaults to 0.5 in the Dropout class constructor.

A slightly more advanced use case is that you want to use different include_probs for some layer. For example, people usually use include_prob=0.8 for the input layer. To do this, you specify include_prob in the constructor arguments for those layers. Other than that, it's the same as the basic use case case. You do nothing special at test time and nothing is dropped. You pass dropout=True at train time and units are dropped based on their input prob.

The most advanced use case is if you want to change dropout include probabilities for a specific fprop call. In this case, you pass dropout=True and dropout_dict for that call. Each layer uses the include_prob specified in the dropout_dict for that call.

Reply:
My comment was mostly wondering about 2. and is more clear now, thanks for making the change! Also there was a typo in my comment, I meant when `include_prob` is 1. (basically disable dropout for one layer only), sorry about that.